### DISCUSSION TITLE
CORS help requested

### DISCUSSION THREAD

**Discussion by mmoore99** [trakt/trakt-api#337] [Q&A] [answered]
I am attempting to access the api from a javascript app running in a browser and am running into CORS problems. The Trakt documentation says:

`When creating your API app, specify the JavaScript (CORS) origins you'll be using. We use these origins to return the headers needed for CORS.`

Can someone provide an example of how to do this or point me toward documentation as to how to accomplish this?

Thanks.

**Comment by mmoore99** [comment]
Further information to the original post:

This is the error message that I get when attempting to access tghe api:

`Access to XMLHttpRequest at 'https://api.trakt.tv/calendars/my/shows' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

I have included the header `origin: http://localhost:3000` in the request but still get the same error.

Is there some other header that needs to be sent with the request?

**Reply by rectifyer** [reply]
When you edit your API app on the Trakt website, there is a text box to set the CORS url. If you add it there, it should start working.

**Reply by mmoore99** [reply]
> When you edit your API app on the Trakt website, there is a text box to set the CORS url. If you add it there, it should start working.

Thanks for pointing that out, sorry I missed it.

Now that the request is working I have another CORS related request. The `access-control-expose-headers` header in the response it set as follows: 
`X-Pagination-Page, X-Pagination-Page-Count, X-Pagination-Limit, X-Pagination-Item-Count, X-Trending-User-Count, X-Start-Date, X-End-Date, X-Sort-By, X-Sort-How, X-Private-User`

Because `x-ratelimit` is not included in the above list the browser does not make it available to the javascript app. Would you consider adding `x-ratelimit` to the `access-control-expose-headers` in the response?

**Reply by rectifyer** [reply]
Great catch. I actually found a few other headers that also needed to be allowed.

`X-Ratelimit, Retry-After, X-Upgrade-URL, X-Applied-Sort-By, X-Applied-Sort-How`

This fix will go out early next week.

**Reply by mmoore99** [reply]
While you are messing with the CORS related header stuff would you also change the `access-control-allow-headers` header to allow `Cache-Control` , `Pragma` and `Expires`. 

When I attempt to send these headers I get CORS errors as well:

`Access to XMLHttpRequest at 'https://api.trakt.tv/sync/last_activities' from origin 'http://localhost:3000' has been blocked by CORS policy: Request header field cache-control is not allowed by Access-Control-Allow-Headers in preflight response.`

**Reply by rectifyer** [reply, marked as answer]
Yes, I'll add those in an update later this week.

**Comment by mmoore99** [comment]
Thanks, I appreciate it!
